plugin/kubernetes/setup.go

```diff
@@ -68,13 +68,19 @@ func (k *Kubernetes) RegisterKubeCache(c *caddy.Controller) {
 		if k.APIProxy != nil {
 			k.APIProxy.Run()
 		}
-		synced := false
-		for synced == false {
-			synced = k.APIConn.HasSynced()
-			time.Sleep(100 * time.Millisecond)
-		}
 
-		return nil
+		timeout := time.After(5 * time.Second)
+		ticker := time.NewTicker(100 * time.Millisecond)
+		for {
+			select {
+			case <-ticker.C:
+				if k.APIConn.HasSynced() {
+					return nil
+				}
+			case <-timeout:
```

Comment:
are we reconnecting to the API when it's available, or are we just dead in the water now?

Reply:
The API connection works in the background trying to do the sync. This just changes the behaviour, of blocking the startup indefinitely waiting for the initial sync (HasSynced) to be true by only waiting 5 seconds for the sync then letting startup continue. 

When the initial sync completes (HasSynced = true) ServeDNS will stop returning ServerFailure for kubernetes zones.